apps/admin/backend/src/cvr_files.ts

```diff
@@ -96,3 +115,282 @@ export async function listCastVoteRecordFilesOnUsb(
+  electionDefinition: ElectionDefinition;
+  reportBallotImageLocations: string[];
+  reportBatchIds: string[];
+}): Result<void, CastVoteRecordValidationError> {
+  const { election, electionHash } = electionDefinition;
+
+  if (cvr.ElectionId !== electionHash) {
+    return err('invalid-election');
+  }
+
+  if (
+    !election.ballotStyles.some(
+      (ballotStyle) => ballotStyle.id === cvr.BallotStyleId
+    )
+  ) {
+    return err('invalid-ballot-style');
+  }
+  const ballotStyle = getBallotStyle({
+    ballotStyleId: cvr.BallotStyleId,
+    election,
+  });
+  assert(ballotStyle);
+
+  if (
+    !election.precincts.some(
+      (precinct) => precinct.id === cvr.BallotStyleUnitId
+    )
+  ) {
+    return err('invalid-precinct');
+  }
+
+  if (!reportBatchIds.some((reportBatchId) => reportBatchId === cvr.BatchId)) {
+    return err('invalid-batch');
+  }
+
+  if (cvr.BallotSheetId) {
+    // Only applicable to hand-marked paper ballots
+    const parseBallotSheetIdResult = safeParseNumber(cvr.BallotSheetId);
+    if (parseBallotSheetIdResult.isErr()) {
+      // TODO: once layouts are in the election definition, we should check
+      // whether the sheet number is within range for the ballot style
+      return err('invalid-sheet-number');
+    }
+  }
+
+  for (const snapshot of cvr.CVRSnapshot) {
+    const contestValidation = snapshotHasValidContestReferences(
+      snapshot,
+      getContests({ ballotStyle, election })
+    );
+    if (contestValidation.isErr()) {
+      return contestValidation;
+    }
+  }
+
+  const ballotImageLocations = cvr.BallotImage?.map(
+    (imageData) => imageData.Location
+  ).filter((location): location is string => location !== undefined);
+
+  if (ballotImageLocations) {
+    for (const ballotImageLocation of ballotImageLocations) {
+      if (!reportBallotImageLocations.includes(ballotImageLocation)) {
+        return err('invalid-ballot-image-location');
+      }
+    }
+  }
+
+  if (!cvrHasValidWriteInImageReferences(cvr)) {
+    return err('invalid-write-in-image-location');
+  }
+
+  if (
+    !cvr.CVRSnapshot.find(
+      (snapshot) => snapshot['@id'] === cvr.CurrentSnapshotId
+    )
+  ) {
+    return err('no-current-snapshot');
+  }
+
+  return ok();
+}
+
+function cvrBallotTypeToLegacyBallotType(
+  ballotType: CVR.vxBallotType
+): CastVoteRecordBallotType {
+  switch (ballotType) {
+    case CVR.vxBallotType.Absentee:
+      return 'absentee';
+    case CVR.vxBallotType.Precinct:
+      return 'standard';
+    case CVR.vxBallotType.Provisional:
+      return 'provisional';
+    default:
+      throwIllegalValue(ballotType);
+  }
+}
+
+/**
+ * Converts the vote data in the CDF cast vote record into the simple
+ * dictionary of contest ids to contest selection ids that VxAdmin uses
+ * internally as a basis for tallying votes.
+ *
+ * This method assumes that if a contest selection is present, then it an
+ * indication. It ignores the possibility that there are multiple selection
```

Comment:
"This method assumes that if a contest selection is present, then it an indication" is missing a verb between "it" and "an"